File: leetcode/python/870_advantage_shuffle.py
class Solution:
    def advantageCount(self, A, B):
        n = len(A)
        order = sorted(range(n), key=lambda x: B[x], reverse=True)
        ans = [-1] * n
        A = deque(sorted(A))
        for ix in order:
            ans[ix] = A.pop() if A[-1] > B[ix] else A.popleft()
        return ans

    def advantageCountAlt(self, A, B):
        ans = []
        A.sort()

        for num in B:
            val = bisect_right(A, num)
            ans.append(A.pop(0) if val == len(A) else A.pop(val))

        return ans

File: leetcode/cpp/870_advantage_shuffle.cpp
class Solution {
public:
    vector<int> advantageCount(vector<int>& A, vector<int>& B) {
        int n = A.size();
        vector<int> order(n);
        vector<int> ans(n, -1);

        for (int i = 0; i < n; i++)
            order[i] = i;

        sort(order.begin(), order.end(), [&B](int a, int b) {
            return B[a] > B[b];
        });

        sort(A.begin(), A.end(), greater<int>());

        int i = 0, j = n - 1;
        for (int ix: order)
            ans[ix] = A[i] > B[ix] ? A[i++] : A[j--];

        return ans;
    }
};
